protocol/opcm-v2.md

```diff
@@ -0,0 +1,234 @@
+
+**2. Unified Initialization Pattern**
+
+Both `deploy()` and `upgrade()` use the same contract initialization path:
+- Clear the `initialized` slot (for upgrades)
+- Call the standard `initialize()` function
+- No special upgrade functions that need removal in subsequent releases
+
+This ensures:
+- Single code path for both operations
+- Always release-ready (no post-upgrade cleanup)
+- Easier testing and maintenance
+
+**3. Auto-Generated Integration Code**
+
+Generate the Go glue code between OPCM and op-deployer from Solidity interfaces:
+- Solidity function signatures/ABIs are the source of truth
+- Go code for constructing function calls is automatically generated
+- Developers only need to add new CLI arguments to op-deployer when introducing net-new inputs
+- Reduces knowledge burden—developers don't need to understand integration layer
+
+### Implementation Flow
+
+Both `deploy()` and `upgrade()` follow a unified three-phase flow:
+
+**Phase 1: Chain World Assembly (Load-or-Deploy Pattern)**
+
+Both functions begin by gathering the complete set of contracts for the chain—the "chain world"—which includes SystemConfig, OptimismPortal, AnchorStateRegistry, and all other chain contracts.
+
+The load-or-deploy pattern attempts to load each contract address from an expected source (e.g., a known deployment address for upgrades, or a predictable location). If the contract cannot be found at the expected location, it is deployed instead. This unified pattern means the same chain world loading function serves both deploy and upgrade operations.
+
+**Phase 2: Input Gathering**
+
+After assembling the chain world, both functions gather the configuration inputs needed to properly configure the proxy contracts:
+
+- **Deployments**: All configuration input comes directly from the deploy input parameter
+- **Upgrades**: Only a subset of configuration is provided in the upgrade input; the remaining configuration is gathered automatically from the existing deployed contracts
+
+This asymmetry allows upgrades to specify only what changes, while deployments must provide complete configuration.
+
+**Phase 3: Common Transformation**
+
+Both functions feed into a shared transformation function that takes the proxies and configuration as input and updates/initializes those proxies appropriately. This common transformation path ensures consistent behavior between initial deployments and subsequent upgrades, eliminating the need for deployment-specific vs. upgrade-specific initialization logic.
+
+### Requirements Mapping
+
+- **Deploy new chains**: `deploy()` function
+- **Upgrade existing chains**: `upgrade()` function  
+- **Dynamic game types**: GameConfig array in Config struct
+- **Add game types**: Implicit in `upgrade()` with modified games array
+- **Update prestates**: Implicit in `upgrade()` with modified games array
```

Comment:
What do you think about preserving `addGameType` and `updatePrestates` as convenience functions? We often want to do just those actions (e.g. changing prestate for a hard fork) and it feels error prone and hard to review to pass a full `Config` into upgrade for that. These convenience functions would just gather the other inputs before calling `upgrade`